vm_backtrace.c

```diff
@@ -626,6 +626,33 @@ bt_iter_iseq(void *ptr, const rb_control_frame_t *cfp)
     arg->prev_loc = loc;
 }
 
+static void
+bt_iter_iseq_skip_internal(void *ptr, const rb_control_frame_t *cfp)
+{
+    static const char prefix[] = "<internal:";
+    const size_t prefix_len = sizeof(prefix) - 1;
+
+    struct bt_iter_arg *arg = (struct bt_iter_arg *)ptr;
+    const rb_iseq_t *iseq = cfp->iseq;
+    VALUE file = rb_iseq_path(cfp->iseq);
+    rb_backtrace_location_t *loc = &arg->bt->backtrace[arg->bt->backtrace_size++-1];
+
+    if (strncmp(prefix, RSTRING_PTR(file), prefix_len) != 0) {
+        const VALUE *pc = cfp->pc;
+        loc->type = LOCATION_TYPE_ISEQ;
+        loc->body.iseq.iseq = iseq;
+        loc->body.iseq.lineno.pc = pc;
+        arg->prev_loc = loc;
+        arg->prev_cfp = cfp;
+    } else if (arg->prev_cfp) {
+        loc->type = LOCATION_TYPE_ISEQ;
+        loc->body.iseq.iseq = arg->prev_cfp->iseq;
+        loc->body.iseq.lineno.pc = arg->prev_cfp->pc;
+        arg->prev_loc = loc;
+    }
+    RB_GC_GUARD(file);
+}
```

Comment:
@jeremyevans Could you review since I think you are more familiar with this code than I am?
Especially the function just above this comment.

Reply:
My understanding of this function:

* Only called for iseq frames
* For non-internal iseq frames, same as bt_iter_iseq, but sets `prev_cfp`
* For internal iseq frames, uses `prev_cfp` to find previous non-internal iseq, and sets iseq information to non-internal frame if one exists

I think the implementation of this function is mostly good, but not sufficient.  You also need a replacement for `bt_iter_skip`.  Because `bt_iter_skip` always sets `prev_cfp` if it is an iseq frame, but you only want to set `prev_cfp` for non-internal iseq frames.  Without that, I think you can still have `uplevel` usage still return an internal frame if there are multiple internal frames in a row.  Setting the `prev_cfp` in `bt_iter_iseq_skip_internal` probably doesn't do anything, since the function should only be called once per backtrace in `uplevel` uses, as `uplevel` use only requests a single frame. The `iter_skip` is called for all frames before the `uplevel` frame.

With a replacement for `bt_iter_skip` that only sets prev_cfp for non-internal frames, I think this should work OK for `uplevel`.  It looks weird for non-`uplevel` usage, though.  A cleaner approach would be a separate frame type for internal frames.  I recently removed the IFUNC frame type as it was unused, something like that would have made sense for this (though ifunc iframe was treated more like cfunc frame than iseq frame).

Since this is only used for `uplevel`, and `rb_ec_backtrace_location_ary` is not exposed to the C-API, this approach should be fine.  We can switch to a separate frame type later if the need arises.

Reply:
Thanks for the detailed review.
My understanding of `backtrace_each` is the stack is walked in the order from "main()" to "current frame", right?

> Without that, I think you can still have `uplevel` usage still return an internal frame if there are multiple internal frames in a row.

Good point.

> Setting the `prev_cfp` in `bt_iter_iseq_skip_internal` probably doesn't do anything, since the function should only be called once per backtrace in `uplevel` uses, as `uplevel` use only requests a single frame.

IIRC I saw cases where `start == 0` and so `iter_skip` was never called. I'll check again.

> Since this is only used for `uplevel`, and `rb_ec_backtrace_location_ary` is not exposed to the C-API, this approach should be fine. We can switch to a separate frame type later if the need arises.

Note that I'd like to be able to filter out `eval(code, binding, "<internal:...", line)` too (to automatically filter out RubyGems' require), so if we use a separate frame type, then it should account for that case too.